eip-0022.md

```diff
@@ -0,0 +1,182 @@
+                        else Coll[Byte]()
+  val isCurrencyERG = auctionCurrency.size == 0
+  val getBoxVal = {(b: Box) => {
+     if (isCurrencyERG) b.value
+     else {
+       if (b.tokens.size == 1 && b.tokens(0)._1 == auctionCurrency) b.tokens(0)._2
+       else if (b.tokens.size == 2 && b.tokens(1)._1 == auctionCurrency) b.tokens(1)._2
+       else 0L
+     }
+  }}
+
+  val currBid = getBoxVal(SELF)
+
+  val auctionConfig = CONTEXT.dataInputs(0)
+
+  // auction is not finished, a bid is being placed
+  val isNextBid = {
+     val extendConfig = auctionConfig.R7[Coll[Long]]
+     val extendThreshold = extendConfig.get(0)
+     val extendNum = extendConfig.get(1)
+     // we extend the auction end time by extendNum if the bid is placed near the very end (extendThreshold)
+     val newEndTime = if (endTime - CONTEXT.preHeader.timestamp <= extendThreshold) endTime + extendNum
+                      else endTime
+
+
+     val newSelf = OUTPUTS(0) // new auction box
+     val oldBidRefundBox = OUTPUTS(1) // refund box
+
+     val isUsingBuyItNow = buyItNowAmount != -1 && getBoxVal(newSelf) >= buyItNowAmount
+
+     // as a general check, ERG value can not be less than the current
+     newSelf.value >= SELF.value &&
+     // auction is not finished
+     CONTEXT.preHeader.timestamp < endTime &&
+     // preserve auctioned tokens
+     newSelf.tokens(0) == SELF.tokens(0) &&
+     // correct value and contract for the new box
+     (getBoxVal(newSelf) >= getBoxVal(SELF) + minStep || getBoxVal(SELF) < minBid || isUsingBuyItNow) &&
+     getBoxVal(newSelf) >= minBid &&
+     newSelf.propositionBytes == SELF.propositionBytes &&
+     // shouldn't be able to add tokens - will change the currency from ERG to a worthless token
+     SELF.tokens.size == newSelf.tokens.size &&
+     // currency must be the same
+     // refund the previous bidder
+     oldBidRefundBox.propositionBytes == currBidder &&
+     getBoxVal(oldBidRefundBox) >= currBid &&
+     // preserve the auction config
+     newSelf.R4[Coll[Byte]].get == seller &&
+     // just making sure that the new R5's type is Coll[Byte]
+     newSelf.R5[Coll[Byte]].get.size > 0 &&
+     newSelf.R6[Coll[Long]].get.size == 2 &&
+     newSelf.R6[Coll[Long]].get(0) == minBid &&
+     newSelf.R6[Coll[Long]].get(1) == minStep &&
+     newSelf.R7[Long].get == newEndTime &&
+     newSelf.R8[Long].get == buyItNowAmount &&
+     newSelf.R9[Coll[Byte]] == SELF.R9[Coll[Byte]]
+
+  }
+
+  // either auction has ended due to time or "Buy it now" is being used
+  val isFinishedWithBid = {
+     val winnerBox = OUTPUTS(0)
+     val auctionFeeBox = OUTPUTS(1)
+     val sellerBox = if (OUTPUTS.size < 3) SELF
+                         else OUTPUTS(2)
+     val artistRoyalty = if (OUTPUTS.size < 4) SELF
+                         else OUTPUTS(3)
+     val originalIssuanceBox = if (auctionFeeBox.R4[Box].isDefined) auctionFeeBox.R4[Box].get
+                               else OUTPUTS(OUTPUTS.size - 1)
```

Comment:
Since the original issuance box cannot be in outputs I guess the reason to get an arbitrary box from the outputs is to have it fail later on id check, right? Would having `val originalIssuanceBox = auctionFeeBox.R4[Box]` at option be more readable?

Reply:
Yes, that would be better. However, this is a trick to get around how the costing currently works. Since both `if` and `else` parts will be evaluated (sometimes, it's weird) that wouldn't work and would produce exceptions.